<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Загрузка-и-проверка-данных" data-toc-modified-id="Загрузка-и-проверка-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Загрузка и проверка данных</a></span></li><li><span><a href="#Обучение-линейной-регрессии" data-toc-modified-id="Обучение-линейной-регрессии-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Обучение линейной регрессии</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Функция-шифрования" data-toc-modified-id="Функция-шифрования-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Функция шифрования</a></span></li><li><span><a href="#Функция-дешефровки" data-toc-modified-id="Функция-дешефровки-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Функция дешефровки</a></span></li><li><span><a href="#Проверка-функцийи-шифровка-признаков" data-toc-modified-id="Проверка-функцийи-шифровка-признаков-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Проверка функцийи шифровка признаков</a></span></li><li><span><a href="#Обучение-модели-на-шифрованных-данных" data-toc-modified-id="Обучение-модели-на-шифрованных-данных-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Обучение модели на шифрованных данных</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

### Загрузка и проверка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
pd.options.mode.chained_assignment = None
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
df = pd.read_csv('/insurance.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
Пол                  5000 non-null int64
Возраст              5000 non-null float64
Зарплата             5000 non-null float64
Члены семьи          5000 non-null int64
Страховые выплаты    5000 non-null int64
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Загрузили данные, проверим что там.
</div>


In [4]:
df.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

In [5]:
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
5 тысяч строк без пропусков, выглядит адекватно, давайте обучим Линейную регрессию и замерим R2, что бы в дальнейшем проверить адекватность наших манипуляций.
</div>


### Обучение линейной регрессии

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Делим данные на выборки, а так же на признаки и целевой признак.
</div>

In [6]:
df_train, df_valid = train_test_split(df, test_size=0.25, random_state=12345)

In [7]:
df_train.shape

(3750, 5)

In [8]:
features_train = df_train.drop('Страховые выплаты', axis=1)
target_train = df_train['Страховые выплаты']

In [9]:
features_valid = df_valid.drop('Страховые выплаты', axis=1)
target_valid = df_valid['Страховые выплаты']

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Создаем и обучаем модель, замеряем R2 которая показывает качество предсказания нашей модели, точнее она показывает зависимость, но нам эта метрика подходит.
</div>

In [10]:
model = LinearRegression()

In [11]:
model.fit(features_train, target_train)
predictions = model.predict(features_valid)
r2 = r2_score(target_valid, predictions)

In [12]:
print(r2)

0.435227571270266


<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Есть отправная точка, переходи к следующему пункуту.
</div>

## Умножение матриц

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">


**Ответ:**  Качество линейной регрессии не изменится при умножении на обратимую (невыраженную) матрицу.
<div>

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
    
**Обоснование:**

1) Элементарные преобразования не меняют ранг матрицы (Рнаг - максимальное число линейно независимых строк\столбцов). Доказано Теоремой об инвариантности ранга при элементарных преобразованиях.

2) Любая невырожденная квадратная матрица является элементарной, другими словами, любая невырожденная квадратная матрица эквивалентна единичной матрице того же порядка. Если $ A $ — невырожденная квадратная матрица n-го порядка, то $ {rg}A=n $ . Поэтому, приводя элементарными преобразованиями матрицу $ A $ к простейшему виду, получим единичную матрицу $ \Lambda=E_n $, так как $ {rg}A=\operatorname{rg}\Lambda=n $. Следовательно, матрица $ A $ эквивалентна единичной матрице $ E_n $ и может быть получена из нее в результате конечного числа элементарных преобразований. Это означает, что матрица $ A $ элементарная. Это следствие Теоремы об инвариантности ранга при элементарных преобразованиях.

3) Умножение на обратимую (невыраженную) матрицу эквивалентно обратимому набору элементарных преобразований и не изменит ранг матрицы, другими словами, не изменит линейной зависимости строк\столбцов в матрице признаков. Таким образом линейная регрессия определит линейные зависимости между признаками так же как и на оригинальных признаках, а в формуле предсказания  $ a = Xw $ на ряду с матрицей признаков $ X $ изменится вектор весов $ w $, а результат $ a $ останется неизменным.
</div>

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
    
**Математическое обоснование:**
    
   **1) Расчет целевого признака без домнождения на обратимую матрицу:**
    
    
Чтобы получить вектор предсказаний $ a $ в ленейной регрессии, мы умножаем матрицу $ X $ на вектор $ w $ и прибавляем величину сдвига $ w0 $.
Для сокращения записи изменим обозначения. В матрицу $ X $ добавим столбец, состоящий только из единиц (он идёт нулевым); а параметр $ w0 $ — к вектору $ w $, получим: $$ a = Xw $$
Далее приведем формулу расчета $ w $, она же формула минимизации потерь и решение задачи линейной регресии: $$ w = (X^TX)^{-1}X^Ty $$
Подставим эту формулу в формулу расчета вектора предсказания $$ a = X(X^TX)^{-1}X^Ty $$
    
    
    
   
   **2) Расчет целевого признака с умножением на обратимую матрицу $ A $:**

Формулу расчета вектора предсказания будет иметь вид $$ a = XAw $$

Формула минимизации потерь и решение задачи линейной регресии примет вид: $$ w = ((XA)^T(XA))^{-1}(XA)^Ty $$
    
Раскрываем скобки с транспонированием: $$ w = (A^TX^TXA)^{-1}A^TX^Ty $$
    
Преобразуем матрицу в первых скобках с обратимостью $$ w = (A)^{-1}(X^TX)^{-1}(A^T)^{-1}A^TX^Ty $$

$ (A^T)^{-1}A^T $ можем сократить, как умножение на обратную матрицу, получаем: $$ w = (A)^{-1}(X^TX)^{-1}X^Ty $$
    
Подставим эту формулу в формулу расчета вектора предсказания: $$ a = XA(A)^{-1}(X^TX)^{-1}X^Ty $$

$ A(A)^{-1} $ можем сократить, получаем: $$ a = X(X^TX)^{-1}X^Ty $$ что в свою очередь равно $ a $ без умножения на обратимую мартицу.
    
   **Вывод**
    
Таким образом, путем элементарных преобразований мы получаем, что вектор предсказаний $ a $ расчитывается идентично как при домнажении на обратимую матрицу $ A $, так и без этого домножения.
    
А так же, исходя из формулы расчета вектора весов $ w $: $ w = (A)^{-1}(X^TX)^{-1}X^Ty $ мы видим, что расчет весов при домножении на матрицу $ A $ отличатся от расчета без умножения, только тем что матрица обратная матрице $ A $ умножается на формулу расчета матрицы весов. 
</div>



## Алгоритм преобразования

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">


**Алгоритм**
 
    1) Шифрование признаков происходит умножением матрици признаков на обратимую матрицу.
    2) Дешифровка признаков происходит умножением зашифрованой матрици признаков на матрицу обратную той, что использовалась для шифровки.
</div>


<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">

**Обоснование**

 1) Как обосновано в разделе два, мы можем умножить матрицу наших признаков на обратимую матрицу и при этом не повлиять на качество предсказания линейной регрессии.
    
 2) Шифрующая обратимая матрица генерируется с использованием метода np.random.seed(), параметр этого значения будет ключем шифрования, так как именно от этого параметра будет зависеть какая матрица будет сгенерирована.
    
 3) При дешифровке мы будем обязаны указать парметр метода .seed() который был указан при шифровании. Так мы будем учверены, что получим матрицу обратную той, что использовалась при шифровании. 
</div>


## Проверка алгоритма

### Функция шифрования

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Создадим функцию шифрования марицы признаков, на вход она будет получать признаки и парметр для генерации шифрующей матрицы, а возвращает матрицу шифрованных признаков. Так же реализуем проверку матрицы на обратимость внутри функции.
</div>


In [13]:
def encrypting(features, random_seed):
    np.random.seed(random_seed)
    m_len = len(features.columns)
    invertible_matrix = np.random.rand(m_len, m_len)
    
    try:
        inverted_matrix = np.linalg.inv(invertible_matrix)
            
    except:
        print("Матрица не обратима (выражена), поробуйте изменить ключь шифрования.")
        
    return np.dot(features, invertible_matrix)

### Функция дешефровки

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Следующим шагом реализуем функцию дешифровки, на вход она получает шифрованные признаки? параметр метода Seed для генерации обратной матрици и имена столбцов. Функция возвращает датафрейм с расшифрованными признаками. 
</div>


In [14]:
def decrypting(features, random_seed, columns):
    np.random.seed(random_seed)
    m_len = len(features[0])
    inverted_matrix = np.linalg.inv(np.random.rand(m_len, m_len))
    return pd.DataFrame(np.dot(features, inverted_matrix).round().astype(int),
                        columns=columns)

### Проверка функцийи шифровка признаков

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Шифруем признаки тренировочной и валидационной выборок.
</div>

In [15]:
features_train_encrypted = encrypting(features_train, 12345)
features_valid_encrypted = encrypting(features_valid, 12345)

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Проверяем результат.
</div>

In [16]:
features_train_encrypted

array([[27135.77045937, 23685.26056699, 27109.23284249, 34828.25143246],
       [43157.25458903, 37666.18945652, 43101.35042651, 55393.68057937],
       [30798.23839187, 26880.88556688, 30762.24182939, 39531.26493088],
       ...,
       [33500.34142835, 29239.41337983, 33462.87969226, 42998.05233489],
       [37532.74604428, 32757.37829064, 37482.92607878, 48178.46301811],
       [30566.17758694, 26676.96608011, 30525.08999414, 39233.72519661]])

In [17]:
features_valid_encrypted

array([[29226.12734297, 25509.30371592, 29193.90135559, 37515.14319577],
       [32306.27910576, 28198.85155761, 32275.33141301, 41466.29200195],
       [31552.0403324 , 27538.83418183, 31516.5934554 , 40496.69205572],
       ...,
       [31699.5507598 , 27667.16597719, 31662.27835918, 40686.34069455],
       [30951.21184729, 27014.19165104, 30915.52806443, 39725.68713557],
       [23318.82334203, 20355.31111831, 23301.49048352, 29929.30158079]])

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
Дешефруем тренировочные признаки.
</div>

In [18]:
features_train_decrypted = decrypting(features_train_encrypted, 12345, features_train.columns)

In [19]:
features_train_decrypted.head()

Пол  Возраст  Зарплата  Члены семьи
0    1       43     36200            1
1    1       34     57600            0
2    0       32     41100            1
3    0       36     45100            1
4    0       33     50600            2

### Обучение модели на шифрованных данных

In [20]:
model_n = LinearRegression()

In [21]:
model_n.fit(features_train_encrypted, target_train)
predictions = model_n.predict(features_valid_encrypted)
r2 = r2_score(target_valid, predictions)

In [22]:
print(r2)

0.43522757127067413


0.435227571270266 - R2 метрика без шифрования.

<div style="border-radius: 15px; border: 1px solid grey; padding: 15px;">
    
**Заключение**    
    
Таким образом считаю поставленную задачу выполненной:
    
    - мы показали, что умножение признаков на обратимую матрицу не сказывается на качестве предсказания линейной регрессии (расхождения в мизерных долях обусловлены округлением при расчетах)
    
    - на основании выше сказанного, мы предложили и успешно реализовали алгоритм шифрования и дешефровки данных пользователей.
    
    
</div>